File: src/main/java/so/aporia/utils/files/FilesManager.java
package so.aporia.utils.files;

import aporia.cc.OsManager;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import so.aporia.utils.Logger;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import java.util.zip.ZipOutputStream;

/**
 * FilesManager — обёртка для работы с файлами Aporia.
 *
 * Поддерживаемые форматы:
 *   .apr  — архив (zip) с content.dat + time.json
 *   .zip  — стандартный zip
 *   .java — исходники
 *   .lua  — скрипты
 *   .cbm  — CatBoost модели
 *   .cfg  — конфиги
 *
 * Базовая директория:
 *   Windows : ~/.apr  (скрытая через attrib +s +h)
 *   Linux/macOS : ~/.config/apr
 */
public class FilesManager {

    // ─── Форматы ────────────────────────────────────────────────────────────────

    public enum FileType {
        APR, ZIP, JAVA, LUA, CBM, CFG, UNKNOWN;

        public static FileType fromPath(Path path) {
            String name = path.getFileName().toString().toLowerCase();
            if (name.endsWith(".apr"))  return APR;
            if (name.endsWith(".zip"))  return ZIP;
            if (name.endsWith(".java")) return JAVA;
            if (name.endsWith(".lua"))  return LUA;
            if (name.endsWith(".cbm"))  return CBM;
            if (name.endsWith(".cfg"))  return CFG;
            return UNKNOWN;
        }
    }

    // ─── Gson ────────────────────────────────────────────────────────────────────

    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();

    // ─── Пути ───────────────────────────────────────────────────────────────────

    /**
     * Корневая директория — делегирует OsManager.mainDirectory.
     * Windows: ~/.apr  |  Linux/macOS: ~/.config/apr
     */
    public static final Path ROOT = OsManager.mainDirectory;

    /** config.apr */
    public static final Path CONFIG_FILE = ROOT.resolve("config.apr");

    /** accounts.apr */
    public static final Path ACCOUNTS_FILE = ROOT.resolve("accounts.apr");

    // ─── Инициализация ──────────────────────────────────────────────────────────

    /**
     * Создаёт ROOT директорию, скрывает её на Windows,
     * затем загружает все impl-файлы.
     */
    public static void init() throws IOException {
        if (!Files.exists(ROOT)) {
            Files.createDirectories(ROOT);
            Logger.success("Created root directory: " + ROOT);
            if (OsManager.getPlatform() == OsManager.Platform.WINDOWS) {
                hideWindowsDirectory(ROOT);
            }
        }

        // загружаем все impl
        so.aporia.utils.files.impl.ChatFile.load();
    }

    private static void hideWindowsDirectory(Path path) {
        try {
            Runtime.getRuntime().exec(new String[]{
                "attrib", "+s", "+h", path.toAbsolutePath().toString()
            });
            Logger.debug("Hidden directory: " + path);
        } catch (IOException e) {
            Logger.warn("Failed to hide directory: " + path);
        }
    }

    // ─── Парсинг путей ──────────────────────────────────────────────────────────

    /**
     * Резолвит путь относительно ROOT если он не абсолютный.
     */
    public static Path resolve(String pathStr) {
        Path p = Paths.get(pathStr);
        return p.isAbsolute() ? p : ROOT.resolve(p);
    }

    /**
     * Резолвит путь с заменой ~ на home пользователя.
     */
    public static Path resolveTilde(String pathStr) {
        if (pathStr.startsWith("~")) {
            pathStr = OsManager.userHome.toString() + pathStr.substring(1);
        }
        return Paths.get(pathStr);
    }

    // ─── .apr формат ────────────────────────────────────────────────────────────

    /**
     * Создаёт .apr файл из произвольного контента (байты).
     * Внутри zip: content.dat + time.json
     */
    public static void writeApr(Path aprPath, byte[] content) throws IOException {
        ensureParentDirs(aprPath);
        // читаем старый created ДО того как перезаписываем файл
        String now = OsManager.getCurrentDateTimeIso();
        String created = now;
        if (Files.exists(aprPath)) {
            try {
                String old = readAprTimeJson(aprPath).getOrDefault("created", now);
                if (!old.isEmpty()) created = old;
            } catch (Exception ignored) {}
        }
        try (ZipOutputStream zos = new ZipOutputStream(new FileOutputStream(aprPath.toFile()))) {
            zos.putNextEntry(new ZipEntry("content.dat"));
            zos.write(content);
            zos.closeEntry();

            zos.putNextEntry(new ZipEntry("time.json"));
            zos.write(buildTimeJson(created, now).getBytes(StandardCharsets.UTF_8));
            zos.closeEntry();
        }
    }

    /**
     * Создаёт .apr файл из строки (UTF-8).
     */
    public static void writeApr(Path aprPath, String content) throws IOException {
        writeApr(aprPath, content.getBytes(StandardCharsets.UTF_8));
    }

    /**
     * Читает content.dat из .apr файла как байты.
     */
    public static byte[] readAprBytes(Path aprPath) throws IOException {
        try (ZipInputStream zis = new ZipInputStream(new FileInputStream(aprPath.toFile()))) {
            ZipEntry entry;
            while ((entry = zis.getNextEntry()) != null) {
                if (entry.getName().equals("content.dat")) {
                    return zis.readAllBytes();
                }
            }
        }
        throw new IOException("content.dat not found in " + aprPath);
    }

    /**
     * Читает content.dat из .apr файла как строку (UTF-8).
     */
    public static String readApr(Path aprPath) throws IOException {
        return new String(readAprBytes(aprPath), StandardCharsets.UTF_8);
    }

    /**
     * Читает time.json из .apr файла, возвращает Map с "created" и "modified".
     */
    public static Map<String, String> readAprTimeJson(Path aprPath) throws IOException {
        Map<String, String> result = new HashMap<>();
        try (ZipInputStream zis = new ZipInputStream(new FileInputStream(aprPath.toFile()))) {
            ZipEntry entry;
            while ((entry = zis.getNextEntry()) != null) {
                if (entry.getName().equals("time.json")) {
                    String json = new String(zis.readAllBytes(), StandardCharsets.UTF_8);
                    // простой парсинг без зависимостей
                    result.put("created",  extractJsonValue(json, "created"));
                    result.put("modified", extractJsonValue(json, "modified"));
                    return result;
                }
            }
        }
        return result;
    }

    private static String buildTimeJson(String created, String modified) {
        return "{\n  \"created\": \"" + created + "\",\n  \"modified\": \"" + modified + "\"\n}";
    }

    // ─── Общие файловые операции ────────────────────────────────────────────────

    /** Читает любой файл как байты. */
    public static byte[] readBytes(Path path) throws IOException {
        return Files.readAllBytes(path);
    }

    /** Читает любой файл как строку (UTF-8). */
    public static String readText(Path path) throws IOException {
        return new String(Files.readAllBytes(path), StandardCharsets.UTF_8);
    }

    /** Записывает байты в файл. */
    public static void writeBytes(Path path, byte[] data) throws IOException {
        ensureParentDirs(path);
        Files.write(path, data);
    }

    /** Записывает строку в файл (UTF-8). */
    public static void writeText(Path path, String text) throws IOException {
        writeBytes(path, text.getBytes(StandardCharsets.UTF_8));
    }

    /** Удаляет файл если существует. */
    public static boolean delete(Path path) {
        try {
            return Files.deleteIfExists(path);
        } catch (IOException e) {
            return false;
        }
    }

    /** Проверяет существование файла. */
    public static boolean exists(Path path) {
        return Files.exists(path);
    }

    /** Возвращает размер файла в байтах, -1 если не существует. */
    public static long size(Path path) {
        try {
            return Files.size(path);
        } catch (IOException e) {
            return -1;
        }
    }

    /** Копирует файл. */
    public static void copy(Path src, Path dst) throws IOException {
        ensureParentDirs(dst);
        Files.copy(src, dst, StandardCopyOption.REPLACE_EXISTING);
    }

    /** Перемещает файл. */
    public static void move(Path src, Path dst) throws IOException {
        ensureParentDirs(dst);
        Files.move(src, dst, StandardCopyOption.REPLACE_EXISTING);
    }

    // ─── .cfg ───────────────────────────────────────────────────────────────────

    /**
     * Читает .cfg как Map<String, String>.
     * Формат: key=value, строки с # — комментарии.
     */
    public static Map<String, String> readCfg(Path path) throws IOException {
        Map<String, String> map = new HashMap<>();
        for (String line : Files.readAllLines(path, StandardCharsets.UTF_8)) {
            line = line.trim();
            if (line.isEmpty() || line.startsWith("#")) continue;
            int eq = line.indexOf('=');
            if (eq > 0) {
                map.put(line.substring(0, eq).trim(), line.substring(eq + 1).trim());
            }
        }
        return map;
    }

    /**
     * Записывает Map<String, String> в .cfg файл.
     */
    public static void writeCfg(Path path, Map<String, String> data) throws IOException {
        StringBuilder sb = new StringBuilder();
        sb.append("# Aporia config\n");
        for (Map.Entry<String, String> e : data.entrySet()) {
            sb.append(e.getKey()).append("=").append(e.getValue()).append("\n");
        }
        writeText(path, sb.toString());
    }

    // ─── Директории ─────────────────────────────────────────────────────────────

    /**
     * Список файлов в директории с фильтром по расширению.
     * extension — например ".apr", ".cfg". Если null — все файлы.
     */
    public static List<Path> listFiles(Path dir, String extension) throws IOException {
        try (Stream<Path> stream = Files.list(dir)) {
            return stream
                .filter(Files::isRegularFile)
                .filter(p -> extension == null || p.getFileName().toString().endsWith(extension))
                .collect(Collectors.toList());
        }
    }

    // ─── Append ─────────────────────────────────────────────────────────────────

    /** Дописывает текст в конец файла (UTF-8). */
    public static void append(Path path, String text) throws IOException {
        ensureParentDirs(path);
        Files.write(path, text.getBytes(StandardCharsets.UTF_8),
            StandardOpenOption.CREATE, StandardOpenOption.APPEND);
    }

    // ─── JSON ────────────────────────────────────────────────────────────────────

    /** Сериализует объект в JSON и записывает в файл. */
    public static void writeJson(Path path, Object obj) throws IOException {
        writeText(path, GSON.toJson(obj));
    }

    /** Читает JSON файл и десериализует в указанный тип. */
    public static <T> T readJson(Path path, Class<T> type) throws IOException {
        return GSON.fromJson(readText(path), type);
    }

    // ─── Утилиты ────────────────────────────────────────────────────────────────

    private static void ensureParentDirs(Path path) throws IOException {
        Path parent = path.getParent();
        if (parent != null && !Files.exists(parent)) {
            Files.createDirectories(parent);
        }
    }

    /** Минимальный JSON value extractor без зависимостей. */
    private static String extractJsonValue(String json, String key) {
        String search = "\"" + key + "\"";
        int idx = json.indexOf(search);
        if (idx < 0) return "";
        int colon = json.indexOf(':', idx + search.length());
        if (colon < 0) return "";
        int start = json.indexOf('"', colon + 1);
        if (start < 0) return "";
        int end = json.indexOf('"', start + 1);
        if (end < 0) return "";
        return json.substring(start + 1, end);
    }

    /**
     * Возвращает FileType по пути.
     */
    public static FileType getFileType(Path path) {
        return FileType.fromPath(path);
    }
}

File: src/so/aporia/utils/files/FilesManager.kt
package so.aporia.utils.files

import aporia.cc.OsManager
import com.chaos.annotation.Obfuscate
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import so.aporia.utils.files.impl.ChatFile
import so.aporia.utils.files.impl.ConfigFile
import so.aporia.utils.user.logger.Logger
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.*
import java.util.*
import java.util.stream.Collectors
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream

@Obfuscate
object FilesManager {

    enum class FileType {
        APR, ZIP, JAVA, LUA, CBM, CFG, UNKNOWN;

        companion object {
            @JvmStatic
            fun fromPath(path: Path): FileType {
                val name = path.fileName.toString().lowercase()
                return when {
                    name.endsWith(".apr") -> APR
                    name.endsWith(".zip") -> ZIP
                    name.endsWith(".java") -> JAVA
                    name.endsWith(".lua") -> LUA
                    name.endsWith(".cbm") -> CBM
                    name.endsWith(".cfg") -> CFG
                    else -> UNKNOWN
                }
            }
        }
    }

    @JvmField
    val GSON: Gson = GsonBuilder().setPrettyPrinting().create()

    @JvmField
    val ROOT: Path = OsManager.mainDirectory

    @JvmField
    val CONFIG_FILE: Path = ROOT.resolve("config.apr")

    @JvmField
    val ACCOUNTS_FILE: Path = ROOT.resolve("accounts.apr")

    @JvmStatic
    @Throws(IOException::class)
    fun init() {
        if (!Files.exists(ROOT)) {
            Files.createDirectories(ROOT)
            Logger.success("Created root directory: $ROOT")
            if (OsManager.platform == OsManager.Platform.WINDOWS) {
                hideWindowsDirectory(ROOT)
            }
        }
        ChatFile.load()
        ConfigFile.load()
    }

    private fun hideWindowsDirectory(path: Path) {
        try {
            Runtime.getRuntime().exec(arrayOf("attrib", "+s", "+h", path.toAbsolutePath().toString()))
            Logger.debug("Hidden directory: $path")
        } catch (e: IOException) {
            Logger.warn("Failed to hide directory: $path")
        }
    }

    @JvmStatic
    fun resolve(pathStr: String): Path {
        val p = Paths.get(pathStr)
        return if (p.isAbsolute) p else ROOT.resolve(p)
    }

    @JvmStatic
    fun resolveTilde(pathStr: String): Path {
        var s = pathStr
        if (s.startsWith("~")) {
            s = OsManager.userHome.toString() + s.substring(1)
        }
        return Paths.get(s)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun writeApr(aprPath: Path, content: ByteArray) {
        ensureParentDirs(aprPath)
        val now = OsManager.getCurrentDateTimeIso()
        var created = now
        if (Files.exists(aprPath)) {
            try {
                val old = readAprTimeJson(aprPath).getOrDefault("created", now)
                if (old.isNotEmpty()) created = old
            } catch (_: Exception) {}
        }
        ZipOutputStream(FileOutputStream(aprPath.toFile())).use { zos ->
            zos.putNextEntry(ZipEntry("content.dat"))
            zos.write(content)
            zos.closeEntry()

            zos.putNextEntry(ZipEntry("time.json"))
            zos.write(buildTimeJson(created, now).toByteArray(StandardCharsets.UTF_8))
            zos.closeEntry()
        }
    }

    @JvmStatic
    @Throws(IOException::class)
    fun writeApr(aprPath: Path, content: String) {
        writeApr(aprPath, content.toByteArray(StandardCharsets.UTF_8))
    }

    @JvmStatic
    @Throws(IOException::class)
    fun readAprBytes(aprPath: Path): ByteArray {
        ZipInputStream(FileInputStream(aprPath.toFile())).use { zis ->
            var entry: ZipEntry?
            while (zis.nextEntry.also { entry = it } != null) {
                if (entry!!.name == "content.dat") {
                    return zis.readAllBytes()
                }
            }
        }
        throw IOException("content.dat not found in $aprPath")
    }

    @JvmStatic
    @Throws(IOException::class)
    fun readApr(aprPath: Path): String {
        return String(readAprBytes(aprPath), StandardCharsets.UTF_8)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun readAprTimeJson(aprPath: Path): Map<String, String> {
        val result = mutableMapOf<String, String>()
        ZipInputStream(FileInputStream(aprPath.toFile())).use { zis ->
            var entry: ZipEntry?
            while (zis.nextEntry.also { entry = it } != null) {
                if (entry!!.name == "time.json") {
                    val json = String(zis.readAllBytes(), StandardCharsets.UTF_8)
                    result["created"] = extractJsonValue(json, "created")
                    result["modified"] = extractJsonValue(json, "modified")
                    return result
                }
            }
        }
        return result
    }

    private fun buildTimeJson(created: String, modified: String): String {
        return "{\n  \"created\": \"$created\",\n  \"modified\": \"$modified\"\n}"
    }

    @JvmStatic
    @Throws(IOException::class)
    fun readBytes(path: Path): ByteArray = Files.readAllBytes(path)

    @JvmStatic
    @Throws(IOException::class)
    fun readText(path: Path): String = String(Files.readAllBytes(path), StandardCharsets.UTF_8)

    @JvmStatic
    @Throws(IOException::class)
    fun writeBytes(path: Path, data: ByteArray) {
        ensureParentDirs(path)
        Files.write(path, data)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun writeText(path: Path, text: String) {
        writeBytes(path, text.toByteArray(StandardCharsets.UTF_8))
    }

    @JvmStatic
    fun delete(path: Path): Boolean {
        return try {
            Files.deleteIfExists(path)
        } catch (_: IOException) {
            false
        }
    }

    @JvmStatic
    fun exists(path: Path): Boolean = Files.exists(path)

    @JvmStatic
    fun size(path: Path): Long {
        return try {
            Files.size(path)
        } catch (_: IOException) {
            -1
        }
    }

    @JvmStatic
    @Throws(IOException::class)
    fun copy(src: Path, dst: Path) {
        ensureParentDirs(dst)
        Files.copy(src, dst, StandardCopyOption.REPLACE_EXISTING)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun move(src: Path, dst: Path) {
        ensureParentDirs(dst)
        Files.move(src, dst, StandardCopyOption.REPLACE_EXISTING)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun readCfg(path: Path): Map<String, String> {
        val map = mutableMapOf<String, String>()
        for (line in Files.readAllLines(path, StandardCharsets.UTF_8)) {
            val trimmed = line.trim()
            if (trimmed.isEmpty() || trimmed.startsWith("#")) continue
            val eq = trimmed.indexOf('=')
            if (eq > 0) {
                map[trimmed.substring(0, eq).trim()] = trimmed.substring(eq + 1).trim()
            }
        }
        return map
    }

    @JvmStatic
    @Throws(IOException::class)
    fun writeCfg(path: Path, data: Map<String, String>) {
        val sb = StringBuilder()
        sb.appendLine("# Aporia config")
        for ((key, value) in data) {
            sb.append(key).append("=").append(value).appendLine()
        }
        writeText(path, sb.toString())
    }

    @JvmStatic
    @Throws(IOException::class)
    fun listFiles(dir: Path, extension: String?): List<Path> {
        Files.list(dir).use { stream ->
            return stream
                .filter { Files.isRegularFile(it) }
                .filter { extension == null || it.fileName.toString().endsWith(extension) }
                .collect(Collectors.toList())
        }
    }

    @JvmStatic
    @Throws(IOException::class)
    fun append(path: Path, text: String) {
        ensureParentDirs(path)
        Files.write(path, text.toByteArray(StandardCharsets.UTF_8), StandardOpenOption.CREATE, StandardOpenOption.APPEND)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun writeJson(path: Path, obj: Any) {
        writeText(path, GSON.toJson(obj))
    }

    @JvmStatic
    @Throws(IOException::class)
    fun <T> readJson(path: Path, type: Class<T>): T {
        return GSON.fromJson(readText(path), type)
    }

    @Throws(IOException::class)
    private fun ensureParentDirs(path: Path) {
        val parent = path.parent
        if (parent != null && !Files.exists(parent)) {
            Files.createDirectories(parent)
        }
    }

    private fun extractJsonValue(json: String, key: String): String {
        val search = "\"$key\""
        val idx = json.indexOf(search)
        if (idx < 0) return ""
        val colon = json.indexOf(':', idx + search.length)
        if (colon < 0) return ""
        val start = json.indexOf('"', colon + 1)
        if (start < 0) return ""
        val end = json.indexOf('"', start + 1)
        if (end < 0) return ""
        return json.substring(start + 1, end)
    }

    @JvmStatic
    fun getFileType(path: Path): FileType = FileType.fromPath(path)
}
